Mathlib/NumberTheory/NumberField/Ideal.lean

```diff
@@ -0,0 +1,155 @@
+
+variable {C : ClassGroup (𝓞 K)} {J : (Ideal (𝓞 K))⁰} {s : ℝ}
+
+private theorem tendsto_norm_le_and_mk_eq_div_atop_aux₁ (hJ : ClassGroup.mk0 J = C⁻¹) :
+    Nat.card {I : (Ideal (𝓞 K))⁰ // absNorm (I : Ideal (𝓞 K)) ≤ s ∧ ClassGroup.mk0 I = C}
+      = Nat.card {I : (Ideal (𝓞 K))⁰ // (J : Ideal (𝓞 K)) ∣ I ∧ IsPrincipal (I : Ideal (𝓞 K)) ∧
+        absNorm (I : Ideal (𝓞 K)) ≤ s * absNorm (J : Ideal (𝓞 K))} := by
+  simp_rw [← nonZeroDivisors_dvd_iff_dvd_coe]
+  refine Nat.card_congr ?_
+  refine ((Equiv.dvd J).subtypeEquiv fun I ↦ ?_).trans
+    (Equiv.subtypeSubtypeEquivSubtypeInter (fun I : (Ideal (𝓞 K))⁰ ↦ J ∣ I) _)
+  rw [← ClassGroup.mk0_eq_one_iff (SetLike.coe_mem _)]
+  simp_rw [Equiv.dvd_apply, Submonoid.coe_mul, ← Submonoid.mul_def, _root_.map_mul, hJ,
+    inv_mul_eq_one, Nat.cast_mul, mul_comm s, eq_comm, and_comm, and_congr_left_iff]
+  exact fun _ ↦
+    (mul_le_mul_iff_of_pos_left (Nat.cast_pos.mpr (absNorm_pos_of_nonZeroDivisors J))).symm
+
+open Classical in
+private def tendsto_norm_le_and_mk_eq_div_atop_aux₂ :
+    ↑({x | x ∈ (toMixed K) ⁻¹' fundamentalCone K ∧ mixedEmbedding.norm ((toMixed K) x) ≤ s} ∩
+      (ZLattice.comap ℝ (idealLattice K ((FractionalIdeal.mk0 K) J)) (toMixed K).toLinearMap))
+        ≃ {a : idealSet K J // mixedEmbedding.norm (a : mixedSpace K) ≤ s} := by
+  rw [ZLattice.coe_comap]
+  refine (((toMixed K).toEquiv.image _).trans (Equiv.setCongr ?_)).trans
+    (Equiv.subtypeSubtypeEquivSubtypeInter _ (mixedEmbedding.norm · ≤ s)).symm
+  ext
+  simp_rw [mem_idealSet, Set.mem_image, Set.mem_inter_iff, Set.mem_preimage, SetLike.mem_coe,
+    mem_idealLattice, FractionalIdeal.coe_mk0]
+  constructor
+  · rintro ⟨_, ⟨⟨hx₁, hx₂⟩, _, ⟨x, hx₃, rfl⟩, rfl⟩, rfl⟩
+    exact ⟨⟨hx₁, x, hx₃, rfl⟩, hx₂⟩
+  · rintro ⟨⟨hx₁, ⟨x, hx₂, rfl⟩⟩, hx₃⟩
+    exact ⟨(toMixed K).symm (mixedEmbedding K x), ⟨⟨hx₁, hx₃⟩, ⟨(x : K), by simp [hx₂], rfl⟩⟩, rfl⟩
+
+variable (C) in
+/--
+The limit of the number of nonzero integral ideals of norm `≤ s` in a fixed class `C` of the
+classgroup divided by `s` when `s → +∞`.
+-/
+theorem tendsto_norm_le_and_mk_eq_div_atop :
+    Tendsto (fun s : ℝ ↦
+      (Nat.card {I : (Ideal (𝓞 K))⁰ //
+        absNorm (I : Ideal (𝓞 K)) ≤ s ∧ ClassGroup.mk0 I = C} : ℝ) / s) atTop
+          (𝓝 ((2 ^ nrRealPlaces K * (2 * π) ^ nrComplexPlaces K * regulator K) /
+            (torsionOrder K *  Real.sqrt |discr K|))) := by
+  classical
+  have h₁ : ∀ s : ℝ,
+    {x | x ∈ toMixed K ⁻¹' fundamentalCone K ∧ mixedEmbedding.norm (toMixed K x) ≤ s} =
+      toMixed K ⁻¹' {x | x ∈ fundamentalCone K ∧ mixedEmbedding.norm x ≤ s} := fun _ ↦ rfl
+  have h₂ : {x | x ∈ fundamentalCone K ∧ mixedEmbedding.norm x ≤ 1} = normLeOne K := by ext; simp
+  obtain ⟨J, hJ⟩ := ClassGroup.mk0_surjective C⁻¹
+  have h₃ : (absNorm J.1 : ℝ) ≠ 0 := (Nat.cast_ne_zero.mpr (absNorm_ne_zero_of_nonZeroDivisors J))
+  convert ((ZLattice.covolume.tendsto_card_le_div'
```

Comment:
Might this proof be clearer the other way around, with `have` statements for each of the hypotheses of `ZLattice.covolume.tendsto_card_le_div'` and then the big theorem at the end, rather than `convert`?